setup.py

```diff
@@ -112,21 +113,42 @@ def __init__(self, name, cmake_path, sources, **kwargs):
         super(CMakeExtension, self).__init__(name, sources=sources, **kwargs)
         self.cmake_path = cmake_path
 
+class FrameworkBuilder(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def cmake_flags(self):
+        return NotImplemented
 
-ext_modules = []
+    @abc.abstractmethod
+    def initialize_options(self):
+        return NotImplemented
 
-ext_modules.append(
-    CMakeExtension(
-        name="transformer_engine",
-        cmake_path=os.path.join(path, "transformer_engine/common"),
-        sources=[],
-        include_dirs=include_dirs,
-    )
-)
+    @abc.abstractmethod
+    def finalize_options(self):
+        return NotImplemented
 
-if framework in ("all", "pytorch"):
-    ext_modules.append(
-        CUDAExtension(
+    @abc.abstractmethod
+    def run(self):
+        return NotImplemented
+
+class PyTorchBuilder(FrameworkBuilder):
+    def __init__(self, *args, **kwargs) -> None:
+        pytorch_args = copy.deepcopy(args)
+        pytorch_kwargs = copy.deepcopy(kwargs)
+        from torch.utils.cpp_extension import BuildExtension
+        self.pytorch_build_extensions = BuildExtension(*pytorch_args, **pytorch_kwargs)
+
+    def cmake_flags(self):
+        return []
+
+    def initialize_options(self):
+        self.pytorch_build_extensions.initialize_options()
+
+    def finalize_options(self):
+        self.pytorch_build_extensions.finalize_options()
+
+    def run(self):
+        from torch.utils.cpp_extension import CUDAExtension
+        other_ext = CUDAExtension(
```

Comment:
I'm not very familiar with `setuptools`, but I wonder if this is a cause of your build error. This `CUDAExtension` does not have a chance to initialize/finalize options before it is run. I wonder if you need to construct it before being passed into `setuptools.setup`: https://github.com/NVIDIA/TransformerEngine/blob/47f1a2910c628d852b9094ed2dc64f3ab63fe948/setup.py#L378

Reply:
This issue was solved.
Moving `CUDAExtension ` back to `ext_modules` can address the error.
I'm not very familiar with setuptools either. This is the best I can find for now.  :disappointed_relieved: